prov/efa/src/rxr/rxr_msg.c

```diff
@@ -55,157 +55,85 @@
-		max_eager_data_size = rxr_pkt_req_max_data_size(rxr_ep,
-								tx_entry->addr,
-								pkt_type + tagged,
-								tx_entry->fi_flags, 0);
-
-		max_eager_data_size = MIN(max_eager_data_size, rxr_env.efa_max_gdrcopy_msg_size);
-
-		if (tx_entry->total_len <= max_eager_data_size) {
-			return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry,
-						 pkt_type + tagged, 0, 0);
-		}
 
-		if (tx_entry->total_len <= rxr_env.efa_max_gdrcopy_msg_size) {
-			if (tx_entry->total_len <= rxr_env.efa_max_medium_msg_size) {
-				pkt_type = delivery_complete_requested ? RXR_DC_MEDIUM_MSGRTM_PKT : RXR_MEDIUM_MSGRTM_PKT;
-			} else {
-				pkt_type = delivery_complete_requested ? RXR_DC_LONGCTS_MSGRTM_PKT : RXR_LONGCTS_MSGRTM_PKT;
-			}
+	eager_rtm = (delivery_complete_requested) ? RXR_DC_EAGER_MSGRTM_PKT + tagged
+						  : RXR_EAGER_MSGRTM_PKT + tagged;
 
-			return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry,
-						 pkt_type + tagged, 0, 0);
-		}
-	}
+	medium_rtm = (delivery_complete_requested) ? RXR_DC_MEDIUM_MSGRTM_PKT + tagged
+						   : RXR_MEDIUM_MSGRTM_PKT + tagged;
 
-	/* At this point we must use read message protocol for cuda memory.
-	 * However, because read message protocol is an extra feature, we do not know
-	 * whether the receiver supports it.
-	 * The only way we can be sure of that is through the handshake packet
-	 * from the receiver, so here we call rxr_pkt_wait_handshake().
-	 */
-	peer = rxr_ep_get_peer(rxr_ep, tx_entry->addr);
-	assert(peer);
+	longcts_rtm = (delivery_complete_requested) ? RXR_DC_LONGCTS_MSGRTM_PKT + tagged
+						    : RXR_LONGCTS_MSGRTM_PKT + tagged;
 
-	err = rxr_pkt_wait_handshake(rxr_ep, tx_entry->addr, peer);
-	if (OFI_UNLIKELY(err)) {
-		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "waiting for handshake packet failed!\n");
-		return err;
-	}
+	longread_rtm = RXR_LONGREAD_MSGRTM_PKT + tagged;
 
-	assert(peer->flags & RXR_PEER_HANDSHAKE_RECEIVED);
-	if (!efa_peer_support_rdma_read(peer)) {
-		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "Cannot send gpu data because receiver does not support RDMA\n");
-		return -FI_EOPNOTSUPP;
+	if (tx_entry->total_len == 0)
+		return eager_rtm;
+
+	if (cuda_is_gdrcopy_enabled() && tx_entry->total_len <= rxr_env.efa_max_gdrcopy_msg_size) {
+		eager_rtm_max_data_size = rxr_pkt_req_max_data_size(rxr_ep, tx_entry->addr, eager_rtm,
+								    tx_entry->fi_flags, 0);
+
+		eager_rtm_max_data_size = MIN(eager_rtm_max_data_size, rxr_env.efa_max_gdrcopy_msg_size);
+
+		if (tx_entry->total_len <= eager_rtm_max_data_size)
+			return eager_rtm;
+
+		if (tx_entry->total_len <= rxr_env.efa_max_medium_msg_size)
+			return medium_rtm;
+
+		return longcts_rtm;
 	}
 
-	return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry,
-				 RXR_LONGREAD_MSGRTM_PKT + tagged, 0, 0);
+	return longread_rtm;
 }
 
-ssize_t rxr_msg_post_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_entry)
+int rxr_msg_select_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_entry, int use_p2p)
```

Comment:
Could u add a docstring for this function?

Reply:
Added in new revision.